File: ecobertura.ui/src/ecobertura/ui/launching/JavaApplicationLaunchShortcut.java
package ecobertura.ui.launching;

import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IConfigurationElement;
import org.eclipse.core.runtime.IExecutableExtension;
import org.eclipse.debug.ui.ILaunchShortcut;
import org.eclipse.jface.viewers.ISelection;
import org.eclipse.ui.IEditorPart;

public class JavaApplicationLaunchShortcut implements ILaunchShortcut, IExecutableExtension {

	private ILaunchShortcut shortcutToExtend;
	private boolean launchDelegationPossible;
	
	@Override
	public void setInitializationData(IConfigurationElement config, String propertyName, 
			Object data) throws CoreException {
		
		shortcutToExtend = LaunchShortcutRetriever.fromId((String) data);
		launchDelegationPossible = (shortcutToExtend != null); 
	}

	@Override
	public void launch(ISelection selection, String mode) {
		if (launchDelegationPossible) {
			shortcutToExtend.launch(selection, mode);
		}
	}

	@Override
	public void launch(IEditorPart editor, String mode) {
		if (launchDelegationPossible) {
			shortcutToExtend.launch(editor, mode);
		}
	}
}

File: ecobertura.ui/src/main/scala/ecobertura/ui/launching/JavaApplicationLaunchShortcut.scala
/*
 * This file is part of eCobertura.
 * 
 * Copyright (c) 2009, 2010 Joachim Hofer
 * All rights reserved.
 *
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */
package ecobertura.ui.launching

import org.eclipse.core.runtime._
import org.eclipse.debug.ui.ILaunchShortcut
import org.eclipse.jface.viewers.ISelection
import org.eclipse.ui.IEditorPart

class JavaApplicationLaunchShortcut extends ILaunchShortcut with IExecutableExtension {
	private var shortcutToExtend : Option[ILaunchShortcut] = None
	
	override def setInitializationData(config: IConfigurationElement, propertyName: String, 
			data: Any) = {
		val retrievedShortcut = LaunchShortcutRetriever.fromId(data.asInstanceOf[String]) 
		if (retrievedShortcut != null) 
			shortcutToExtend = Some(retrievedShortcut) 
	}
	
	override def launch(selection: ISelection, mode: String) = shortcutToExtend match {
		case Some(shortcut) => shortcut.launch(selection, mode)
		case _ => // do nothing
	}
	
	override def launch(editor: IEditorPart, mode: String) = shortcutToExtend match {
		case Some(shortcut) => shortcut.launch(editor, mode)
		case _ => // do nothing
	}
}
